### PR TITLE
Typed library

### PR DESCRIPTION
**Pull request by k0t3n**
* Add strict typings for `theine` lib
* Deprecate python3.17, add python3.12 tests

### CODE REVIEW

**Comment by k0t3n** [comment]
@Yiling-J don't you mind making the library fully typed? I submitted an example for the Django adapter, please have a look. If it's ok for you, I can fix the typings for the rest of the codebase.

**Comment by Yiling-J** [comment]
@k0t3n That looks really good! Feel free to add others.

**Comment by Yiling-J** [comment]
Hi @k0t3n I plan to start refactoring/rewriting to support free-threaded in Python 3.13 (https://github.com/Yiling-J/theine/issues/28). Since this might involve a significant refactor/rewrite, I think it would be better to wait your type improvements and releasing a new version for that. Do you have an estimate for when this will be finished?

**Comment by k0t3n** [comment]
@Yiling-J I have almost finished with theine lib, but some work still needs to be done on `tests/` and `benchmarks`/, which is tricky in some places. I still wanna finish it up, but unfortunately cannot give you any estimations until September.

BWT can we drop support for python 3.7 within these changes? It's not compatible with some stubs and it's EOL finished more than a year ago.

**Comment by Yiling-J** [comment]
@k0t3n I think it's fine to drop 3.7 support. My concern is that if I make significant changes to `theine.py`, there will be lots  conflicts. However I can use your branch as a starting point if you're almost finished with theine lib files.

File: theine/theine.py [hunk 1]
```diff
@@ -8,53 +8,54 @@
...
+from mypy_extensions import KwArg, VarArg
 from theine_core import ClockProCore, LruCore, TlfuCore
 from typing_extensions import ParamSpec, Protocol
 
 from theine.exceptions import InvalidTTL
 from theine.models import CacheStats
 
+P = ParamSpec("P")
+R = TypeVar("R", covariant=True, bound=Any)
+R_A = TypeVar("R_A", covariant=True, bound=Union[Awaitable[Any], Callable[..., Any]])
```
Comment (Yiling-J) [inline theine/theine.py:31]: seems R is replaced by R_A? but R_A bound to `Union[Awaitable[Any], Callable[..., Any]]`, so this code
```python
@Memoize(Cache("tlfu", 1000), None)
def foo(id: int, m: Mock) -> Dict[str, int]:
    m(id)
    return {"id": id}
```
will not pass mypy check. I already tried and this is the error:
```
error: Value of type variable "R_A" of "__call__" of "Memoize" cannot be "dict[str, int]"  [type-var]
```

**Comment by Yiling-J** [comment]
@k0t3n I think it might be better to add a separate test/example file for all public APIs and have mypy check that file in CI. The actual tests/benchmarks don't need to be typed, as they sometimes need to access internal properties, and typing them could make things more complicated.

**Pull request merged**